File: app/src/main/java/com/bytesforge/linkasanote/synclog/SyncLogAdapter.java
/*
 * LaaNo Android application
 *
 * @author Aleksandr Borisenko <developer@laano.net>
 * Copyright (C) 2017 Aleksandr Borisenko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.bytesforge.linkasanote.synclog;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.bytesforge.linkasanote.data.SyncResult;
import com.bytesforge.linkasanote.databinding.ItemSyncLogBinding;

import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;

public class SyncLogAdapter extends RecyclerView.Adapter<SyncLogAdapter.ViewHolder> {

    private static final String TAG = SyncLogAdapter.class.getSimpleName();

    private final SyncLogViewModel viewModel;

    private List<SyncResult> syncResults;

    public SyncLogAdapter(
            @NonNull List<SyncResult> syncResults,
            @NonNull SyncLogViewModel viewModel) {
        this.syncResults = checkNotNull(syncResults);
        this.viewModel = checkNotNull(viewModel);
        setHasStableIds(true);
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        private final ItemSyncLogBinding binding;

        public ViewHolder(ItemSyncLogBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(
                int position, boolean showStarted,
                SyncResult syncResult, SyncLogViewModel viewModel) {
            binding.setPosition(position);
            binding.setShowStarted(showStarted);
            binding.setSyncResult(syncResult);
            binding.setViewModel(viewModel); // NOTE: global viewModel for fragment and all items

            binding.executePendingBindings();
        }
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        ItemSyncLogBinding binding = ItemSyncLogBinding.inflate(inflater, parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        SyncResult syncResult = syncResults.get(position);
        boolean showStarted = false;
        if (position <= 0) {
            showStarted = true;
        } else {
            SyncResult prevSyncResult = syncResults.get(position - 1);
            if (syncResult.getStarted() != prevSyncResult.getStarted()) {
                showStarted = true;
            }
        }
        holder.bind(position, showStarted, syncResult, viewModel);
    }

    @Override
    public int getItemCount() {
        return syncResults.size();
    }

    @Override
    public long getItemId(int position) {
        return syncResults.get(position).getRowId();
    }

    // Items

    @NonNull
    public List<SyncResult> getSyncResults() {
        return syncResults;
    }

    public void swapItems(@NonNull List<SyncResult> syncResults) {
        checkNotNull(syncResults);
        this.syncResults = syncResults;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/bytesforge/linkasanote/synclog/SyncLogAdapter.kt
/*
 * LaaNo Android application
 *
 * @author Aleksandr Borisenko <developer@laano.net>
 * Copyright (C) 2017 Aleksandr Borisenko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.bytesforge.linkasanote.synclog

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bytesforge.linkasanote.data.SyncResult
import com.bytesforge.linkasanote.databinding.ItemSyncLogBinding

class SyncLogAdapter(
    syncResults: List<SyncResult>,
    viewModel: SyncLogViewModel
) : RecyclerView.Adapter<SyncLogAdapter.ViewHolder>() {
    private val viewModel: SyncLogViewModel

    // Items
    var syncResults: List<SyncResult>
        private set

    class ViewHolder(private val binding: ItemSyncLogBinding) : RecyclerView.ViewHolder(
        binding.root
    ) {
        fun bind(
            position: Int, showStarted: Boolean,
            syncResult: SyncResult?, viewModel: SyncLogViewModel?
        ) {
            binding.position = position
            binding.showStarted = showStarted
            binding.syncResult = syncResult
            binding.viewModel = viewModel // NOTE: global viewModel for fragment and all items
            binding.executePendingBindings()
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = ItemSyncLogBinding.inflate(inflater, parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val syncResult = syncResults[position]
        var showStarted = false
        if (position <= 0) {
            showStarted = true
        } else {
            val prevSyncResult = syncResults[position - 1]
            if (syncResult.started != prevSyncResult.started) {
                showStarted = true
            }
        }
        holder.bind(position, showStarted, syncResult, viewModel)
    }

    override fun getItemCount(): Int {
        return syncResults.size
    }

    override fun getItemId(position: Int): Long {
        return syncResults[position].rowId
    }

    fun swapItems(syncResults: List<SyncResult>) {
        this.syncResults = syncResults
        notifyDataSetChanged()
    }

    companion object {
        private val TAG = SyncLogAdapter::class.java.simpleName
        private val TAG_E = SyncLogAdapter::class.java.canonicalName
    }

    init {
        this.syncResults = syncResults
        this.viewModel = viewModel
        setHasStableIds(true)
    }
}